bundles/io/org.eclipse.smarthome.io.voice/src/main/java/org/eclipse/smarthome/io/voice/text/AbstractRuleBasedInterpreter.java

```diff
@@ -446,28 +513,53 @@ protected String executeSingle(ResourceBundle language, String[] nameFragments,
 
     /**
      * Filters the item registry by matching each item's name with the provided name fragments.
-     * For this the item's name is at first tokenized by {@link splitName}.
-     * The resulting fragments are now looked up by each and every provided fragment.
-     * For the item to get included into the result list, every provided fragment has to be found among the item's ones.
+     * The item's label and its parent group's labels are tokenizend {@link tokenize} and then altogether looked up
+     * by each and every provided fragment.
+     * For the item to get included into the result list, every provided fragment has to be found among the label
+     * tokens.
      * If a command type is provided, the item also has to support it.
+     * In case of channels and their owners being disambiguous due to sharing most of the label sequence, only the top
+     * most item with support for the
+     * given command type is kept.
      *
-     * @param nameFragments name fragments that are used to match an item's name.
-     *            For a positive match, the item's name has to contain every fragment - independently of their order.
+     * @param language Language information that is used for matching
+     * @param labelFragments label fragments that are used to match an item's label.
+     *            For a positive match, the item's label has to contain every fragment - independently of their order.
      *            They are treated case insensitive.
      * @param commandType optional command type that all items have to support.
      *            Provide {null} if there is no need for a certain command to be supported.
      * @return All matching items from the item registry.
      */
-    protected ArrayList<Item> getMatchingItems(String[] nameFragments, Class<?> commandType) {
+    protected ArrayList<Item> getMatchingItems(ResourceBundle language, String[] labelFragments, Class<?> commandType) {
         ArrayList<Item> items = new ArrayList<Item>();
-        for (Item item : itemRegistry.getAll()) {
-            HashSet<String> parts = new HashSet<String>(splitName(item.getLabel(), true));
-            boolean allMatch = true;
-            for (String fragment : nameFragments) {
-                allMatch = allMatch && parts.contains(fragment.toLowerCase());
-            }
-            if (allMatch && (commandType == null || item.getAcceptedCommandTypes().contains(commandType))) {
-                items.add(item);
+        HashMap<Item, ArrayList<HashSet<String>>> map = getItemTokens(language.getLocale());
+        for (Item item : map.keySet()) {
+            for (HashSet<String> parts : map.get(item)) {
+                boolean allMatch = true;
+                for (String fragment : labelFragments) {
+                    allMatch = allMatch && parts.contains(fragment.toLowerCase(language.getLocale()));
```

Comment:
I'd use an if statement here instead of && because the code will be clearer and you can break out of the loop early when you find the first non-match.